lib/ReorderBasicBlocksPass.cpp

```diff
@@ -41,21 +54,94 @@ namespace clspv {
 FunctionPass *createReorderBasicBlocksPass() {
   return new ReorderBasicBlocksPass();
 }
+} // namespace clspv
+
+void ReorderBasicBlocksPass::StructuredOrder(BasicBlock *block,
+                                             const LoopInfo &LI,
+                                             std::deque<BasicBlock *> *order,
+                                             DenseSet<BasicBlock *> *visited) {
+  if (!visited->insert(block).second)
+    return;
+
+  // Identify the merge and continue blocks for special treatment.
+  const auto *terminator = dyn_cast<TerminatorInst>(block->getTerminator());
+  BasicBlock *continue_block = nullptr;
+  BasicBlock *merge_block = nullptr;
+  if (LI.isLoopHeader(block)) {
+    Loop *loop = LI.getLoopFor(block);
+    merge_block = loop->getExitBlock();
+
+    if (loop->isLoopLatch(block)) {
+      continue_block = block;
+    } else {
+      DominatorTree &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+      auto *header = loop->getHeader();
+      auto *latch = loop->getLoopLatch();
+      for (auto *bb : loop->blocks()) {
+        if (bb == header)
+          continue;
+
+        if (DT.dominates(bb, latch))
+          continue_block = bb;
+      }
+    }
+    assert(continue_block && merge_block && "Bad loop");
+  } else if (terminator->getNumSuccessors() > 1) {
+    bool has_back_edge = false;
```

Comment:
// This is potentially a selection case.  But it could also be a conditional branch with one arm pointing back to the loop header, which would make this block the latch block for the loop.